clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java

```diff
@@ -1253,26 +1248,48 @@ protected void updateOffsets(
   @Override
   protected void updateOffsetMetadataInOffsetRecord(
       PartitionConsumptionState partitionConsumptionState,
-      OffsetRecord offsetRecord,
-      ConsumerRecord<KafkaKey, KafkaMessageEnvelope> consumerRecordWrapper,
-      LeaderProducedRecordContext leaderProducedRecordContext,
-      String kafkaUrl) {
-    updateOffsets(
+      ConsumerRecord<KafkaKey, KafkaMessageEnvelope> consumerRecord,
+      String upstreamKafkaUrl) {
+    updateOffsetsFromPartitionConsumptionState(
         partitionConsumptionState,
-        consumerRecordWrapper,
-        leaderProducedRecordContext,
-        offsetRecord::setCheckpointLocalVersionTopicOffset,
-        (sourceKafkaUrl, upstreamTopicName, upstreamTopicOffset) -> {
-          if (Version.isRealTimeTopic(upstreamTopicName)) {
-            offsetRecord.setLeaderUpstreamOffset(sourceKafkaUrl, upstreamTopicOffset);
-          } else {
-            offsetRecord.setCheckpointUpstreamVersionTopicOffset(upstreamTopicOffset);
-          }
-        },
-        (sourceKafkaUrl, upstreamTopicName) -> Version.isRealTimeTopic(upstreamTopicName)
-            ? offsetRecord.getUpstreamOffset(sourceKafkaUrl)
-            : offsetRecord.getCheckpointUpstreamVersionTopicOffset(),
-        () -> OffsetRecord.NON_AA_REPLICATION_UPSTREAM_OFFSET_MAP_KEY);
+        () -> OffsetRecord.NON_AA_REPLICATION_UPSTREAM_OFFSET_MAP_KEY,
+        shouldUpdateUpstreamOffset(consumerRecord));
+  }
+
+  /**
+   * Sync the metadata about offset from {@link PartitionConsumptionState} to {@link OffsetRecord}.
+   * It's agnostic to the record as {@link PartitionConsumptionState} should have handled it already.
+   * @param partitionConsumptionState
+   * @param sourceKafkaUrl, The computed upstream Kafka URL for the record. Can be null for shutdown events.
+   * @param shouldUpstreamOffset, the flag to update the upstream offset or not. If the record contains invalid upstream
+   *                              offset, then upstream kafka url doesn't need to be computed unless it's a shutdown event
+   */
+  protected void updateOffsetsFromPartitionConsumptionState(
+      PartitionConsumptionState partitionConsumptionState,
+      Supplier<String> sourceKafkaUrl,
+      boolean shouldUpstreamOffset) {
+    OffsetRecord offsetRecord = partitionConsumptionState.getOffsetRecord();
+    offsetRecord
+        .setCheckpointLocalVersionTopicOffset(partitionConsumptionState.getLatestProcessedLocalVersionTopicOffset());
+    // DaVinci clients don't need to maintain leader production states
+    if (!isDaVinciClient) {
+      if (sourceKafkaUrl == null) {
+        // kafka url is not provided, meaning it's a graceful shutdown event
+        // so ignore the flag and copy the upstream offset map from pcs.
+        offsetRecord.resetUpstreamOffsetMap(partitionConsumptionState.getLatestProcessedUpstreamRTOffsetMap());
```

Comment:
One concern I had for this line is the `partitionState.leaderOffset` in the `OffsetRecord` is no longer valid since the entire map is synced up with PCS. I can probably iterate the map to find out the max upstream offset and use it for `partitionState.leaderOffset`. 

From this [old commit](https://github.com/linkedin/venice/commit/63ea4435fa27fc0f4d55382a808bb6864bd0eca5), turned out we might have deprecated/is deprecating this field. If so, it may be okay to not update this when reset the map.

@huangminchn @shuhui-liu , can you please double-check it for me?

Reply:
Would it be the case that even looping and finding the max can not guarantee the consistency, because `resetUpstreamOffsetMap` doesn't perform a deep copy of the object, but a simple pointer assignment. Any changes to the PartitionConsumptionState::latestProcessedUpstreamRTOffsetMap later could still be seen in the persistent data?

Reply:
That's a good point. By the time of shutdown, we will wait for the complete processing (including persistence to disk) of all the messages and update PCS, prior to calling this `syncMetadataInOffsetRecord` method, so there should have no more updates in the PCS.

However, I think we can make a deep copy as defensive coding just in case

Reply:
Could copy the map from pcs to offsetrecord by looping and calling method `OffsetRecord.setLeaderUpstreamOffset(String upstreamKafkaURL, long leaderOffset)` which will update both `upstreamOffsetMap` and `leaderOffset`.